runtime/src/iree/modules/vmvx/module.c

```diff
@@ -61,13 +61,121 @@ iree_vmvx_module_free_state(void* self, iree_vm_module_state_t* module_state) {
 }
 
 //===----------------------------------------------------------------------===//
-// TODO
+// Exported function definitions
 //===----------------------------------------------------------------------===//
 
-// Placeholder to make the function pointer arrays happy (they can't be empty).
-IREE_VM_ABI_EXPORT(iree_vmvx_module_placeholder,  //
-                   iree_vmvx_module_state_t,      //
-                   v, v) {
+static iree_device_size_t iree_vmvx_cast_device_size(int64_t value) {
+  // TODO(benvanik): make this return status and check for overflow if device
+  // size is 32-bits.
+  return (iree_device_size_t)value;
+}
+
+IREE_VM_ABI_EXPORT(iree_vmvx_add_2d_f32, iree_vmvx_module_state_t,
+                   rIIIrIIIrIIIII, v) {
+  // LHS
+  iree_vm_buffer_t* lhs_buffer;
+  IREE_RETURN_IF_ERROR(iree_vm_buffer_check_deref(args->r0, &lhs_buffer));
+  iree_device_size_t lhs_offset = iree_vmvx_cast_device_size(args->i1);
+  iree_device_size_t lhs_stride0 = iree_vmvx_cast_device_size(args->i2);
+  iree_device_size_t lhs_stride1 = iree_vmvx_cast_device_size(args->i3);
+  // RHS
+  iree_vm_buffer_t* rhs_buffer;
+  IREE_RETURN_IF_ERROR(iree_vm_buffer_check_deref(args->r4, &rhs_buffer));
+  iree_device_size_t rhs_offset = iree_vmvx_cast_device_size(args->i5);
+  iree_device_size_t rhs_stride0 = iree_vmvx_cast_device_size(args->i6);
+  iree_device_size_t rhs_stride1 = iree_vmvx_cast_device_size(args->i7);
+  // OUT
+  iree_vm_buffer_t* out_buffer;
+  IREE_RETURN_IF_ERROR(iree_vm_buffer_check_deref(args->r8, &out_buffer));
+  iree_device_size_t out_offset = iree_vmvx_cast_device_size(args->i9);
+  iree_device_size_t out_stride0 = iree_vmvx_cast_device_size(args->i10);
+  iree_device_size_t out_stride1 = iree_vmvx_cast_device_size(args->i11);
+  // SIZE
+  iree_device_size_t size0 = iree_vmvx_cast_device_size(args->i12);
+  iree_device_size_t size1 = iree_vmvx_cast_device_size(args->i13);
+
+  float* lhs = (float*)lhs_buffer->data.data + lhs_offset;
```

Comment:
we'll need a range check helper for offset/size/etc (could stub it out today)

Reply:
https://github.com/google/iree/blob/main/runtime/src/iree/vm/buffer.c#L18-L70 does this today - we could expose iree_vm_buffer_map_ro and iree_vm_buffer_map_rw in the .h as it'd be helpful in all these methods for taking care of things (will ensure we don't try to write into read-only memory, etc)

Reply:
I gave it a shot. ptal. I don't like it...